### PR TITLE
 AC_Avoid: Convert Dijkstras to A-star

### PR DESCRIPTION
**Pull request by rishabsingh3003**
TLDR: This converts present Dijkstra's to A*. This guarantees an optimal path, just like Dijkstra. The number of nodes explored to find the shortest path is a fraction of what is being done in master

This is built on top of: https://github.com/ArduPilot/ardupilot/pull/18806

My testing map has been: 
image

Currently, in master, ALL the nodes are visited and the shortest path is calculated.. this means, the algorithm will consider 27 points + orgin+ goal node every time = 29 nodes every time 

In my above PR, this is reduced to: 
image

It's okay... but we can do better!
This PR does this: 
A*

**In many cases, the number of nodes considered (and therefore the time taken to calculate the path) is almost 5-10% of what it takes in master!**

**Because the heuristics is admissible ( I haven't done a mathematical proof, but seems trivial) , this guarantees an optimal path and therefore (at least to me), it sounds like there is no downside to this PR and we should completely replace Dijkstra's with this.**

### CODE REVIEW

**Comment by rishabsingh3003** [comment]
I haven't replaced the 130 odd "Dijkstra" (class name, file name, parameter name, macros etc) because it'll mask and hide away what is actually a very simple PR. It will also make it tough to review.. I'll do that later if needed

**Comment by IamPete1** [comment]
It would be nice to keep path planning flexible so we could use a different cost functions. I would have to do some reading about A* it could be fine. We might want to replace the current distance cost function with time/fuel/altitude/soaring/sailing cost functions.

**Comment by rishabsingh3003** [comment]
@IamPete1 that certainly sounds like a good idea! Although I must say, with cost functions, and especially with Heuristics, one has to be very very careful while coming up with them. Even a slightly wrong heuristics can plan a path miles away from what it should... literally speaking. 
Having said that, I have been getting very well acquainted with the Dijkstra's code, and there are several places we could optimize to make things much faster than what it is. And perhaps we should refactor the code in a way to keep in mind that distance isn't the only cost function criteria.

**Comment by hendjoshsr71** [comment]
Ha ha I like the idea of about the custom cost function! But of course one must be sure it is admissible. My thought on that was adding a cost for changes in momentum. That could reduce the amount of time spent slowing down or changing directions. 

My other thought **NOT** for this PR is that we should strive towards a more generalized approach that allows an easier replacement of the planner type: search method and graph creation.

Such an approach would allow us to do unit testing easier on say this A* method and I think more importantly allow for researchers and new developers to be more comfortable trying new things and contributing back. 

For example for the unit test feeding the A* method would take in the set of obstacles, current location, and destination. Output of where to go next.  We can check that is the optimal path.

**Comment by peterbarker** [comment]
On Sat, 2 Oct 2021, Josh Henderson wrote:

> My other thought NOT for this PR is that we should strive towards a more generalized approach that allows an easier replacement of the planner type: search method and graph creation.

Uh - we already have a frontend/backend split for path planning.

Having a common base class (or some other method of code-sharing) that 
would allow A* and Dijkstras to co-exist with minimal overhead would be 
nice.

**Review by rmackay9 — APPROVED** [review]
LGTM, I'll merge after my meeting later today with Rishabh.
